pkg/sql/opt/indexrec/index_recommendation_set.go

```diff
@@ -23,81 +23,141 @@ import (
 	case *memo.InvertedJoinExpr:
-		irs.addIndexToRecommendationSet(expr.Index, expr.Table)
+		irs.addIndexToRecommendationSet(expr.Index, expr.Cols, expr.Table)
 	case *memo.ZigzagJoinExpr:
-		irs.addIndexToRecommendationSet(expr.LeftIndex, expr.LeftTable)
-		irs.addIndexToRecommendationSet(expr.RightIndex, expr.RightTable)
+		irs.addIndexToRecommendationSet(expr.LeftIndex, expr.Cols, expr.LeftTable)
+		irs.addIndexToRecommendationSet(expr.RightIndex, expr.Cols, expr.RightTable)
 	}
 	for i, n := 0, expr.ChildCount(); i < n; i++ {
 		irs.createIndexRecommendations(expr.Child(i))
 	}
 }
 
+// pruneIndexRecommendations removes redundant index recommendations from the
+// recommendation set. These are recommendations where the existingIndex field
+// is not nil and no new columns are being stored.
+func (irs *IndexRecommendationSet) pruneIndexRecommendations() {
+	for t, indexRecs := range irs.indexRecs {
+		updatedIndexRecs := make([]indexRecommendation, 0, len(indexRecs))
+		for _, indexRec := range indexRecs {
+			if indexRec.existingIndex == nil || !indexRec.redundantRecommendation() {
+				updatedIndexRecs = append(updatedIndexRecs, indexRec)
+			}
+		}
+		irs.indexRecs[t] = updatedIndexRecs
+	}
+}
+
+// getColOrdSet returns an ordinal set of an opt.ColSet, containing all columns
+// in the set from the table with id tabID.
+func (irs *IndexRecommendationSet) getColOrdSet(
+	cols opt.ColSet, tabID opt.TableID,
+) util.FastIntSet {
+	var colsOrdSet util.FastIntSet
+	for _, col := range cols.ToList() {
+		// Do not add unknown columns.
+		if col == 0 {
+			continue
+		}
+		table := irs.md.ColumnMeta(col).Table
+		// Do not add columns from other tables.
+		if table != tabID {
+			continue
+		}
+		colOrd := table.ColumnOrdinal(col)
+		colsOrdSet.Add(colOrd)
+	}
+	return colsOrdSet
+}
+
 // addIndexToRecommendationSet adds an index to the indexes map if it does not
 // exist already in the map and in the table.
 func (irs *IndexRecommendationSet) addIndexToRecommendationSet(
-	indexOrd cat.IndexOrdinal, tabID opt.TableID,
+	indexOrd cat.IndexOrdinal, usedCols opt.ColSet, tabID opt.TableID,
 ) {
+	// Do not recommend the primary index.
+	if indexOrd == cat.PrimaryIndex {
+		return
+	}
+	// Do not add real table indexes (non-hypothetical table indexes).
 	switch hypTable := irs.md.TableMeta(tabID).Table.(type) {
 	case *hypotheticalTable:
-		// Do not add real table indexes (non-hypothetical indexes).
-		if indexOrd < hypTable.Table.IndexCount() {
-			return
+		usedColsOrds := irs.getColOrdSet(usedCols, tabID)
+		// Try to find an identical existing index recommendation.
+		for _, indexRec := range irs.indexRecs[hypTable] {
+			index := indexRec.index
+			if index.indexOrdinal == indexOrd {
+				// Update newStoredColOrds to include all stored column ordinals that
+				// are in usedColsOrds.
+				usedStoredColOrds := index.storedColsOrdSet.Intersection(usedColsOrds)
+				indexRec.newStoredColOrds.UnionWith(usedStoredColOrds)
```

Comment:
nit: this logic could be contained in a method for indexRec, like `indexRec.AddStoredColOrds(scannedColOrds)`